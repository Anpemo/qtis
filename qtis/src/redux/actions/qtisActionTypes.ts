export default {
  USER_REGISTER: 'USER_REGISTER',
  PRODUCTS_LIST: 'PRODUCTS_LIST',
  SINGLE_PRODUCT: 'SINGLE_PRODUCT',
  CREATE_PRODUCT: 'CREATE_PRODUCT',
  REVIEWS_LIST: 'REVIEWS_LIST'
}
