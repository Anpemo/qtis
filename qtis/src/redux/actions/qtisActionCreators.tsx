import qtisActionTypes from './qtisActionTypes'
import axios from 'axios'
import serverUrls from '../../../constants/serverUrls'
import UserInterface from '../../Interfaces/UserInterface'
import ProductInterface from '../../Interfaces/ProductInterface'
import ReviewInterface from '../../Interfaces/ReviewInterface'
import { Dispatch } from 'react'
import { userRegisterAction, updateUserAction, createReviewAction, userLoginAction, fetchProductsAction, fetchProductAction, fetchReviewsAction, createProductAction } from '../../models'

export function userRegister (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<userRegisterAction>) {
    try {
      const { data } = await axios.post(serverUrls.BACKEND_REGISTER, userData)

      dispatch({
        type: qtisActionTypes.USER_REGISTER,
        data
      })
    } catch {
      dispatch({
        type: qtisActionTypes.USER_REGISTER,
        data: 401
      })
    }
  }
}

export function userLogin (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<userLoginAction>) {
    try {
      const { data } = await axios.post(serverUrls.BACKEND_LOGIN, userData)
      dispatch({
        type: qtisActionTypes.USER_LOGIN,
        data
      })
    } catch {
      dispatch({
        type: qtisActionTypes.USER_LOGIN,
        data: 401
      })
    }
  }
}

export function fetchProducts (category: Object) {
  return async function fetchInfo (dispatch: Dispatch<fetchProductsAction>) {
    const { data } = await axios.get(`${serverUrls.PRODUCTS}${category}`)
    dispatch({
      type: qtisActionTypes.PRODUCTS_LIST,
      data
    })
  }
}

export function fetchProduct (barCodeData: String) {
  return async function fetchInfo (dispatch: Dispatch<fetchProductAction>) {
    let date
    try {
      const { data, headers } = await axios.get(`${serverUrls.PRODUCTS}${barCodeData}`)
      date = headers.date
      dispatch({
        type: qtisActionTypes.SINGLE_PRODUCT,
        data: data[0]
      })
    } catch {
      dispatch({
        type: qtisActionTypes.SINGLE_PRODUCT,
        data: { data: 'Product does not exist', date }
      })
    }
  }
}

export function createProduct (productData: ProductInterface) {
  return async function fetchInfo (dispatch: Dispatch<createProductAction>) {
    const { data } = await axios.post(serverUrls.PRODUCTS, productData)
    dispatch({
      type: qtisActionTypes.CREATE_PRODUCT,
      data
    })
  }
}

export function cleanProduct () {
  return ({
    type: qtisActionTypes.CLEAN_PRODUCT,
    data: null
  })
}

export function fetchReviews (parameter: Object) {
  return async function fetchInfo (dispatch: Dispatch<fetchReviewsAction>) {
    const { data } = await axios.get(`${serverUrls.REVIEWS}${parameter}`)

    dispatch({
      type: qtisActionTypes.REVIEWS_LIST,
      data
    })
  }
}

export function createReview (reviewData: ReviewInterface) {
  return async function fetchInfo (dispatch: Dispatch<createReviewAction>) {
    const { data } = await axios.post(serverUrls.REVIEWS, reviewData)
    dispatch({
      type: qtisActionTypes.CREATE_REVIEW,
      data
    })
  }
}

export function updateUser (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<updateUserAction>) {
    const { data } = await axios.put(serverUrls.UPDATE_USER, userData)
    dispatch({
      type: qtisActionTypes.UPDATE_USER,
      data
    })
  }
}
