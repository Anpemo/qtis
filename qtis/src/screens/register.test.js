import { cleanup, fireEvent, render } from '@testing-library/react-native'
import Register from './register'
import React from 'react'
import { Provider } from 'react-redux'
import configureStore from '../redux/stores/configureStore'
import * as actions from '../redux/actions/qtisActionCreators'
jest.mock('../redux/actions/qtisActionCreators')

describe('Given a register component', () => {
  afterEach(() => {
    cleanup()
  })

  describe('When writing on inputs', () => {
    let component
    beforeEach(() => {
      const store = configureStore
      component = (
          <Provider store={store}><Register /></Provider>
      )
    })
    test('Then userName will change', () => {
      const { getByPlaceholderText } = render(component)

      const userNameInput = getByPlaceholderText('What\'s your name?')
      const newValue = 'newValue'
      fireEvent.changeText(userNameInput, newValue)
      expect(userNameInput.props.value).toBe(newValue)
    })
    test('Then email will change', () => {
      const { getByPlaceholderText } = render(component)

      const emailInput = getByPlaceholderText('Email')
      const newValue = 'alegnapm@gmail.com'
      fireEvent.changeText(emailInput, newValue)
      expect(emailInput.props.value).toBe(newValue)
    })
    test('Then password will change', () => {
      const { getByPlaceholderText } = render(component)

      const passwordInput = getByPlaceholderText('Password')
      const newValue = 'a1234567'
      fireEvent.changeText(passwordInput, newValue)
      expect(passwordInput.props.value).toBe(newValue)
    })
    test('Then confirmpassword will change', () => {
      const { getByPlaceholderText } = render(component)
      const confirmpasswordInput = getByPlaceholderText('Confirm Pasword')
      const newValue = 'a1234567'
      fireEvent.changeText(confirmpasswordInput, newValue)
      expect(confirmpasswordInput.props.value).toBe(newValue)
    })
  })
  describe('When pressing on back button', () => {
    test('Then it will navigate', () => {
      const store = configureStore
      const goBack = jest.fn()
      const component = (
          <Provider store={store}><Register navigation={{ goBack }}/></Provider>
      )
      const { getByTestId } = render(component)
      const backButton = getByTestId('backButton')
      fireEvent.press(backButton)
      expect(goBack).toHaveBeenCalled()
    })
  })
  describe('When pressing the registe onPress', () => {
    test('Then it will call userRegister with email', () => {
      jest.spyOn(actions, 'userRegister').mockReturnValue({ type: '' })
      const store = configureStore
      const component = (
          <Provider store={store}><Register /></Provider>
      )
      const { getByText } = render(component)
      fireEvent.press(getByText('REGISTER'))
      expect(actions.userRegister).toHaveBeenCalled()
    })
  })
})
