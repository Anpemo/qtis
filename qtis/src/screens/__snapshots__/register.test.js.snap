// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a register component When pressing the registe onPress Then it will call userRegister with email 1`] = `[MockFunction]`;
